arch/x86/compare258_avx512.c

```diff
@@ -0,0 +1,84 @@
+/* compare258_avx512.c -- AVX512 version of compare258
+ * Copyright Adam Stylinski  <kungfujesus06@gmail.com>
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "fallback_builtins.h"
+
+#if defined(X86_AVX512) && defined(HAVE_BUILTIN_CTZ)
+
+/* Some versions of clang and GCC forgot to add these. Without them sometimes extra
+ * trips to GPRs are taken, but this is otherwise compatible */
+#ifndef X86_MASK_INTRIN
+#define _knot_mask64(a) ~a
+#endif
+
+#include <immintrin.h>
+#ifdef _MSC_VER
+#  include <nmmintrin.h>
+#endif
+
+/* UNALIGNED_OK, AVX512 intrinsic comparison */
+static inline uint32_t compare256_unaligned_avx512_static(const unsigned char *src0, const unsigned char *src1) {
+    uint32_t len = 0;
+
+    do {
+        __m512i zmm_src0, zmm_src1;
+        __mmask64 mask;
+        zmm_src0 = _mm512_loadu_si512(src0);
+        zmm_src1 = _mm512_loadu_si512(src1);
+        /* non-identical bytes = 00, identical bytes = FF */
+        mask = _mm512_cmp_epi8_mask(zmm_src0, zmm_src1, _MM_CMPINT_EQ);
+        /* Invert it here, since comparing against 0 is cheap and it will need
+         * to be inverted later, anyway */
+        mask = _knot_mask64(mask);
+        if (mask) {
+            uint32_t match_byte = (uint32_t)__builtin_ctzl(mask); /* Invert bits so identical = 0 */
+            return len + match_byte;
+        }
+
+        src0 += 64, src1 += 64, len += 64;
+
+        zmm_src0 = _mm512_loadu_si512(src0);
+        zmm_src1 = _mm512_loadu_si512(src1);
+        mask = _mm512_cmp_epi8_mask(zmm_src0, zmm_src1, _MM_CMPINT_EQ);
+        mask = _knot_mask64(mask);
+        if (mask) {
+            uint32_t match_byte = (uint32_t)__builtin_ctzl(mask);
```

Comment:
__builtin_ctzl is not available on MSVC. Maybe use __builtin_ctzll instead? 

Or add __builtin_ctzl to fallback functions..

Reply:
ll should work, I think.  Silly distinctions these days, may as well be calling pointers far pointers.